core/commands/dag/dag.go

```diff
@@ -241,3 +277,487 @@ var DagResolveCmd = &cmds.Command{
+				err,
+			)
+		}
+
+		// The current interface of go-car is rather suboptimal as it
+		// only takes a blockstore, instead of accepting a dagservice,
+		// and leveraging parallel-fetch capabilities
+		//
+		// Until the above is fixed, pre-warm the blockstore before doing
+		// anything else. We explicitly *DO NOT* take a lock during this
+		// operation: even if we lose some of the blocks we just received
+		// due to a conflicting GC: we will just re-retrieve anything we
+		// potentially lost when the car is being streamed out
+		node, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		if err := mdag.FetchGraph(req.Context, c, node.DAG); err != nil {
+			if !node.IsOnline {
+				err = fmt.Errorf("%s (currently offline, perhaps retry after attaching to the network)", err)
+			}
+			return err
+		}
+
+		// The second part of the above - make a super-thin wrapper around
+		// a blockservice session, translating Session.GetBlock() to Blockstore.Get()
+		//
+		// sess := blockservice.NewSession(
+		// 	req.Context,
+		// 	node.Blocks,
+		// )
+		// var wrapper getBlockFromSessionWrapper = func(c cid.Cid) (blk.Block, error) {
+		// 	return sess.GetBlock(req.Context, c)
+		// }
+		// sb := gipselectorbuilder.NewSelectorSpecBuilder(gipfree.NodeBuilder())
+		// car := gocar.NewSelectiveCar(
+		// 	req.Context,
+		// 	&wrapper,
+		// 	[]gocar.Dag{gocar.Dag{
+		// 		Root: c,
+		// 		Selector: sb.ExploreRecursive(
+		// 			gipselector.RecursionLimitNone(),
+		// 			sb.ExploreAll(sb.ExploreRecursiveEdge()),
+		// 		).Node(),
+		// 	}},
+		// )
+
+		pipeR, pipeW := io.Pipe()
+
+		errCh := make(chan error, 2) // we only report the 1st error
+		go func() {
+			defer func() {
+				if err := pipeW.Close(); err != nil {
+					errCh <- fmt.Errorf("stream flush failed: %s", err)
+				}
+				close(errCh)
+			}()
+
+			//if err := car.Write(pipeW); err != nil {
+			if err := gocar.WriteCar(
+				req.Context,
+				node.DAG,
+				[]cid.Cid{c},
+				pipeW,
+			); err != nil {
+				errCh <- err
+			}
+		}()
+
+		if err := res.Emit(pipeR); err != nil {
+			pipeW.Close() // ignore errors if any
+			return err
+		}
+
+		return <-errCh
+	},
+	PostRun: cmds.PostRunMap{
+		cmds.CLI: func(res cmds.Response, re cmds.ResponseEmitter) error {
+			if showProgress, _ := res.Request().Options[progressOptionName].(bool); !showProgress {
+				return cmds.Copy(re, res)
+			}
+
+			var exportSize int64
+			buf := make([]byte, 4*1024*1024)
+			defer func() {
+				if exportSize > 0 {
+					os.Stderr.WriteString("\n")
+				}
+			}()
+
+			for {
+				v, err := res.Next()
+				if err != nil {
+					if err == io.EOF {
+						return re.Close()
+					}
+					return re.CloseWithError(err)
+				}
+
+				if r, ok := v.(io.Reader); ok {
+					// we got a reader passed as a response
+					// proxy it through with an increasing counter
+					for {
+						len, readErr := r.Read(buf)
+						if len > 0 {
+							if err := re.Emit(bytes.NewBuffer(buf[:len])); err != nil {
```

Comment:
This may happen to work (should add a check to prevent it) when not running the daemon, but simply won't work when actually running the daemon. When running the daemon, we can only return:

1. Repeated results of the same type (encoded as json). To do this, we have to specify the type.
2. Streaming bytes (no type specified).

We can't return both.

Reply:
I wasn't sure abut both not being possible, so opted for the loop. Will simplify.

Reply:
Sorry, I misread this the first time. We _can_ repeatedly stream new byte readers.

However, we really shouldn't read 4MiB, write 4MiB, etc. Ideally, we'd print out a rate and amount downloaded.

Reply:
We do not read/write 4MiB. This is the max we can read from the pipe at once, but other than that we just block and pull through whatever slice is placed on the pipe.

https://github.com/ipfs/go-ipfs/pull/7037/files#diff-15172926c4147422f472139313467179R418-R421